### ISSUE TITLE
use of flag_values in face_face_connectivity not consistent with CF

### ISSUE THREAD

**Issue by GLang-BAW** [ugrid-conventions/ugrid-conventions#26]
(1) Actual use of flag_values and flag_meanings in connectivity list.

   integer Mesh2_face_links(nMesh2_face, Three) ;
      :cf_role = "face_face_connectivity" ;
      :long_name = "Indicates which other faces neighbor each face." ;
      :start_index = 1 ;
      :flag_values = -1 ;
      :flag_meanings = "out_of_mesh" ;

(2) What CF says about use of flag_values and flag_meanings

   The attributes flag_values and flag_meanings are intended to make variables that contain
   flag values self describing.  The flag_values and flag_meanings attributes describe
   a status flag consisting of mutually exclusive coded values. 

(3) My understanding of variables with flag_values and flag_meanings

   Each data item of the variable is interpreted in terms of the available flag_values, flag_meanings.
   
   For each valid data item there must be an entry in the list of flag_values, flag_meanings. Otherwise
   we have no meaning for a specific data item.
   
   For the above variable I would expect to have a second entry in flag_meanings with e.g. value inside_mesh.

    Conclusion: flag_values and flag_meanings are not used consistent with CF in the current definition
    of the above printed connectivity list.

(4) Recommendation: Use _FillValue to indicate missing neighbor (out of mesh). Example:

   integer Mesh2_face_links(nMesh2_face, Three) ;
      :cf_role = "face_face_connectivity" ;
      :long_name = "Indicates which other faces neighbor each face." ;
      :start_index = 1 ;
      :_FillValue = -1 ;

  This is, from my point of view, also much more simple and straightforward to use. For faces with more than three edges one one might not be able to compute the number of edges for this face based on the face_face_connectivity table alone. For this purpose one should use e.g. face_edge_connectivity.

Günther Lang

**Comment by ChrisBarker-NOAA** [comment]
@GLang-BAW: THanks for notciing this -- I agree, it seems that flag_values is being missused here,

However,  _FillValue is used by netcdf as the "value used to fill an array whos data has not been specified, and there is a default for each datatype -- what is it for integers???

but here:
http://cfconventions.org/cf-conventions/cf-conventions.html#missing-data

is says to use it, rather than `missing_value`, so I"m a bit confused.

Of course, in this case, it's only sort of a missing value, but this does seem to be the best option in CF.

The other alternative is to simply specify in the convention that -1 is used to indicate no neighbor, and that's that -- no attribute to the variable to describe  it at all -- unless we are going to allow applications to use something other than -1 -- I really hope not!

-CHB

**Comment by GLang-BAW** [comment]
1. I suggest to use only the _FillValue attribute in order to indicate that there are missing data,
    i. e. not use the missing_value attribute in addition to _FillValue.

2. Within UGRID we do already use _FillValue as the sole indicator for missing data, e. g. in the
   face_node_connectivity table. So using _FillValue in face_face_connectivity as the sole indicator
   for missing data would be OK for me.

3. In our organisation we do use flexible mesh almost always in combination with data defined on
   faces _and_ edges. Therefore I do also suggest, that an edge_face_connectivity table should be 
   allowed as optional attribute in the mesh variable. Example:

   int Mesh2_edge_faces(nMesh2_edge, two) ;
      :long_name = "edge face connectivity table, left and right neighbours" ;
      :cf_role = "edge_face_connectivity" ;
      :start_index = 0 ;
      :_FillValue = -999 ;
      
   Use of an edge_face_connectivity table has the following advantage compared to face_face_connectivity: For each edge there are always two entries (for faces the number varies from face to face in dependence on the number of nodes/edges, so we can not distinguish between missing edges and missing neighbours from face_face_connectivity alone).

Günther

**Comment by ChrisBarker-NOAA** [comment]
@GLang-BAW wrote:
> I suggest to use only the _FillValue attribute in order to indicate that there are missing data,
> i. e. not use the missing_value attribute in addition to _FillValue.

Sounds good -- but what about simply codifying that -1 means "out of grid" or "no neighbor there" or whatever is appropriate -- then we don't need to bother with setting _FillVAlue -- and, more importantly, file-reading libs don't need to check and use that value.

Please edit the standard and make a PR -- unless someone else speaks up with an objection.

@rsignell-usgs, @hrajagers: any thoughts???

**Comment by ChrisBarker-NOAA** [comment]
I created #27 for (3) -- as it's a separate proposal.

**Comment by arthurvd** [comment]
About specifying which "missing value" for topology variables should be
used:

I would vote for standard use of _FillValue (instead of requiring -1).
Motivation:
* it is the easiest to understand for all CF-aware people and CF-compliant
readers. For example: in Python the NetCDF readers automatically change
each variable's fill values to None (null)
* when no _FillValue attribute is specified, the defaults are available via
the header files: NF90_FILL_INT/NC_FILL_INT (-2147483647L)
* Indeed, we also already use this for existing face_node_connectivity
arrays and the like.
* We could even suggest a default -1, but let's always annotate that in the
variable via the :_FillValue=-1. If not we're suggesting our own standard
for something already wll known in CF.

A mild motivation against it:
* keeping the _FillValue on default -2147483647L, and requiring in UGRID
the value -1 for "no neighbour" would allow distinguishing between 1) not
(yet) written data, e.g. in incompletely written files, and 2) true,
intended absent topological neighbours.
However, I think this is a far-fetched example, and I'd stick to using
_FillValue as motivated above.

Best,
Arthur van Dam

**Comment by ChrisBarker-NOAA** [comment]
@arthurvd wrote:

> * it is the easiest to understand for all CF-aware people and CF-compliant
> readers. For example: in Python the NetCDF readers automatically change
> each variable's fill values to None (null)

ouch! that's actually a good reason NOT to use it -- when I read those arrays, I WANT -1 to be there, NOT None (actually there is no None in numpy arrays, nor NaN in integer arrays, but I think it does produce a masked array.

but this is not really a "missing value" which is what _FillValue is usually interpreted as -- it a a special value that has a special meaning -- which, I suppose is why "flags" was chosen in the first place.

* when no _FillValue attribute is specified, the defaults are available via
the header files: NF90_FILL_INT/NC_FILL_INT (-2147483647L)

and in pythons netCDF4 somewhere, I imagine. -- but I would call it non-standards compliant to jsut use the default FillValue without specifying it.

* Indeed, we also already use this for existing face_node_connectivity
arrays and the like.
-- use the default FillValue? rather than -1 or specifying it?

* We could even suggest a default -1, but let's always annotate that in the
variable via the :_FillValue=-1. If not we're suggesting our own standard
for something already well known in CF.

except this isn't really a Missing value -- it is a specific to this use-case special value.

There is always a trade off in standards between letting people essentially write the files however they are already doing it (or at least the most convenient way), and having everything more standardised. As a data consumer more than a data creator, I personally prefer strict standards -- makes it much easier to write code that reads the data...

But I won't put up any barriers, but if _FillVAlue is auto-filled by the python libs (and any others), I'd rather NOT use it, because my code is now going to have to deal with that :-(

**Comment by arthurvd** [comment]
@ChrisBarker-NOAA wrote:

> but this is not really a "missing value" which is what _FillValue is usually interpreted as -- it a a special value that has a special meaning -- which, I suppose is why "flags" was chosen in the first place.

Ok, this will indeed depend on whether we want to distinguish between "missing" and "special".
For this type of connectivity variables, how would "special" differ from "missing"? I am assuming that a "missing" value means: there is no neighbour cell/extra corner nodes/etc. on this location.

CF has no clear answer for our topology variables, but says something related on auxiliary coordinate variables/locations. In the current CF-1.6 and 1.7 conventions I read:

> Auxiliary coordinate variables (spatial and time) must contain missing values to indicate a void in data storage in the file but must not have missing data for any other reason. This situation may arise for unused elements in the incomplete multidimensional array representation, and in any representation if the instance dimension is set to a larger size than the number of features currently stored.
> (http://cfconventions.org/cf-conventions/cf-conventions.html#_missing_data)


@ChrisBarker-NOAA wrote:
> * Indeed, we also already use this for existing face_node_connectivity
> arrays and the like.
> 
>   -- use the default FillValue? rather than -1 or specifying it?
> 

No, I meant using the concept of _FillValue at all to denote "absent" nodes. In the current UGRID-1.0 I read:

> The connectivity array will be a matrix of size nFaces x MaxNumNodesPerFace; **if a face has less corner nodes than MaxNumNodesPerFace then the last node indices shall be equal to _FillValue** (which should obviously be larger than the number of nodes in the mesh)

Notice how here another suggestion for _FillValue is being done (> num_nodes)

@ChrisBarker-NOAA wrote:
>  As a data consumer more than a data creator, I personally prefer strict standards -- makes it much easier to write code that reads the data...
> 
> But I won't put up any barriers, but if _FillVAlue is auto-filled by the python libs (and any others), I'd rather NOT use it, because my code is now going to have to deal with that :-(
> 

At Deltares we both produce and consume data, and I really wouldn't want to create unnecessary work for any of us consumers out there. But I'm currently not sure why an autofilled mask would be problematic: if there's no neighbouring cells, then I guess nothing has to be done on that position? (I'm probably missing a point here, no pun intended)

**Comment by ChrisBarker-NOAA** [comment]
>> But I'm currently not sure why an autofilled mask would be problematic: if there's no neighbouring cells, then I guess nothing has to be done on that position?

The issue is that masked arrays are not universal -- either outside numpy or even within numpy. So we need to use a special value (which is -1 everywhere I've seen it) to mark the "no-neighbor" condition.

So it would be a touch easier if the -1 was guaranteed to be already there in the array.

**Comment by GLang-BAW** [comment]
If we follow [CF conventions on missing data](http://cfconventions.org/cf-conventions/cf-conventions.html#missing-data) we can read there: "If only one missing value is needed for a variable then we recommend that this value be specified using the _FillValue attribute."

1. In UGRID we should use _FillValue as sole attribute for "missing data";
2. Any reasonable number for _FillValue is allowed to be chosen (it must be a number which is outside the range of valid data, whether positive or negative doesn't matter). Any UGRID compliant software has to deal properly with different values for _FillValue.
3. Flag values should not be used in connectivity tables to indicate special missing data.

Comment: If one wants to provide more information about a variable, one has to create an additional variable which only holds flag values (plus _FillValues where there are no data at all). This new "flag variable" is then listed in the ancillary_variables attribute of the original variable, e.g. connectivity table. But this should not become a required feature in UGRID.

Günther

**Comment by ChrisBarker-NOAA** [comment]
>> No, I meant using the concept of _FillValue at all to denote "absent" nodes. In the current UGRID-1.0 I read:

>> The connectivity array will be a matrix of size nFaces x MaxNumNodesPerFace; if a face has less corner nodes than MaxNumNodesPerFace then the last node indices shall be equal to _FillValue (which should obviously be larger than the number of nodes in the mesh)

>> Notice how here another suggestion for _FillValue is being done (> num_nodes)

interesting -- we might want to add "negative OR larger than the number of nodes in the mesh" -- maybe it's that way to accommodate unsigned types?

But anyway, maybe it's just me, but "this face has only four nodes" (i.e .the fifth and sixth are _FillValue). really does mean "missing value", whereas, this face has six side, but the this one is on the boundary of the mesh, i.e has no neighbor on that side. Are conceptually different.

Though now that I write it, maybe not :-)

**Comment by arthurvd** [comment]
@ChrisBarker-NOAA wrote:

> But anyway, maybe it's just me, but "this face has only four nodes" (i.e .the fifth and sixth are _FillValue). really does mean "missing value", whereas, this face has six side, but the this one is on the boundary of the mesh, i.e has no neighbor on that side. Are conceptually different.

Ha! Good example, because this first brings up a new question (and partly this may also be the reason why @GLang-BAW brought up the request in #27).

Question: In a face_face_connectivity list, are the neighbouring faces order in the same way as the edges that lie in between? (This would be very convenient and useful, but UGRID-1.0 does not specify it.)

In other words: given a grid with only tris and quads, and there exists a tri # 10 whose first edge lies on the grid boundary (no neighbouring face), and whose second and third edge have neighbouring faces # 12 and #14 for example. How does the face_face_connectivity for this triangle look like?
`_, 12, 14, _`
or
`12, 14, _, _`

**Comment by ChrisBarker-NOAA** [comment]
indeed -- nice catch! I haven't dealt with any multi-number-of-edges meshes, so I didn't thikn of that. but do we need to make a distinction between:

This cell only has 3 edges
and
this cell has 4 edges, but the fourth has no neighbor on that side.

In the connectivity arrays?

Also -- I guess we didn't specify it in the spec, but it's certainly common and good practice to keep everything in the same order everywhere. If there is a consistent order, then the concept of second and third exists, and in your example:

"""
In other words: given a grid with only tris and quads, and there exists a tri # 10 whose first edge lies on the grid boundary (no neighbouring face), and whose second and third edge have neighbouring faces # 12 and # 14 for example. How does the face_face_connectivity for this triangle look like?
_, 12, 14, _
or
12, 14, _, _
"""
_, 12, 14, _

As I write this, I think you need to keep a consistent order!

but even so, do we need to have:

no_neighbor, 12, 14, no_edge_at_all

i.e. two different types of flag value.

-CHB
